File: ufa/services/analytical-moose-foobar/app/apis/foo/consumption/foo-filters-values-api.ts
import { ConsumptionApi } from "@514labs/moose-lib";
import { FooPipeline } from "../../../index";
import {
  GetFooFiltersValuesParams,
  GetFooFiltersValuesResponse,
} from "@workspace/models/foo";

export const fooFiltersValuesApi = new ConsumptionApi<
  GetFooFiltersValuesParams,
  GetFooFiltersValuesResponse
>(
  "foo-filters-values",
  async (
    { months = 6 }: GetFooFiltersValuesParams,
    { client, sql }
  ): Promise<GetFooFiltersValuesResponse> => {
    const endDate = new Date();
    const startDate = new Date();
    startDate.setMonth(
      startDate.getMonth() - Math.max(1, Math.min(months, 36))
    );

    const startDateStr = startDate.toISOString().split("T")[0];
    const endDateStr = endDate.toISOString().split("T")[0];

    // Query for distinct status and tag values
    const statusQuery = sql`
      SELECT DISTINCT status
      FROM ${FooPipeline.table!}
      WHERE toDate(created_at) >= toDate(${startDateStr})
        AND toDate(created_at) <= toDate(${endDateStr})
        AND status IS NOT NULL
        AND cdc_operation != 'DELETE'
      ORDER BY status
    `;

    const tagsQuery = sql`
      SELECT DISTINCT arrayJoin(tags) AS tag
      FROM ${FooPipeline.table!}
      WHERE toDate(created_at) >= toDate(${startDateStr})
        AND toDate(created_at) <= toDate(${endDateStr})
        AND tags IS NOT NULL
        AND cdc_operation != 'DELETE'
      ORDER BY tag
    `;

    const [statusSet, tagsSet] = await Promise.all([
      client.query.execute<{ status: string }>(statusQuery),
      client.query.execute<{ tag: string }>(tagsQuery),
    ]);

    const statusRows = (await statusSet.json()) as { status: string }[];
    const tagRows = (await tagsSet.json()) as { tag: string }[];

    const possibleStatuses = statusRows.map((row) => row.status);
    const possibleTags = tagRows.map((row) => row.tag);

    return {
      status: possibleStatuses,
      tags: possibleTags,
    };
  }
);

File: ufa-lite/services/analytical-moose-foobar-py/app/apis/foo/consumption/foo_filters_values_api.py
from moose_lib import Api
from typing import Optional, List
from pydantic import BaseModel
from datetime import datetime, timedelta


class GetFooFiltersValuesParams(BaseModel):
    months: Optional[int] = 6


class GetFooFiltersValuesResponse(BaseModel):
    status: List[str]
    tags: List[str]
    priorities: List[int]


def foo_filters_values_api_handler(
    context,
    params: GetFooFiltersValuesParams
) -> GetFooFiltersValuesResponse:
    """
    API to get distinct filter values for foo data
    """
    # Calculate date range
    end_date = datetime.now()
    start_date = end_date - timedelta(days=max(1, min(params.months, 36)) * 30)

    start_date_str = start_date.strftime("%Y-%m-%d")
    end_date_str = end_date.strftime("%Y-%m-%d")

    # Build queries for distinct values
    date_filter = f"""
        toDate(created_at) >= toDate('{start_date_str}')
        AND toDate(created_at) <= toDate('{end_date_str}')
    """

    status_query = f"""
        SELECT DISTINCT status
        FROM foo
        WHERE {date_filter}
          AND status IS NOT NULL
        ORDER BY status
    """

    tags_query = f"""
        SELECT DISTINCT arrayJoin(tags) AS tag
        FROM foo
        WHERE {date_filter}
          AND tags IS NOT NULL
        ORDER BY tag
    """

    priorities_query = f"""
        SELECT DISTINCT priority
        FROM foo
        WHERE {date_filter}
          AND priority IS NOT NULL
        ORDER BY priority
    """

    # Execute queries
    status_results = context.query(status_query, {})
    tags_results = context.query(tags_query, {})
    priorities_results = context.query(priorities_query, {})

    # Extract values from results
    possible_statuses = [row["status"] for row in status_results]
    possible_tags = [row["tag"] for row in tags_results]
    possible_priorities = [int(row["priority"]) for row in priorities_results]

    return GetFooFiltersValuesResponse(
        status=possible_statuses,
        tags=possible_tags,
        priorities=possible_priorities
    )


# Create the API instance
foo_filters_values_api = Api[GetFooFiltersValuesParams, GetFooFiltersValuesResponse](
    name="foo-filters-values",
    query_function=foo_filters_values_api_handler
)